### ISSUE TITLE
View Encapsulation not working between multiple Angular 4 Apps (built with aot)

### ISSUE THREAD

**Issue by d-bechtel** [angular/angular#16676]
**I'm submitting a ...**  (check one with "x")
```
[X] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

If i boot multiple (different) Angular 4 Apps (in different zones), the styles of each app will be adopt to all apps, because of the same css selector ( .app[_ngcontent-c0] )
![angular_aot_bug](https://cloud.githubusercontent.com/assets/22303562/25885855/1ff17938-355b-11e7-8ec9-60cd1143d805.PNG)

**Expected behavior**

Every Angular 4 app should have its own css-selector

**Minimal reproduction of the problem with instructions**

Here is an working example with JIT-compiled Apps, both apps get different css selectors. If i build these apps with AOT, both Apps get the same css selector:
http://plnkr.co/edit/fFTrK07EWBV6vSxhYaFe?p=preview

**What is the motivation / use case for changing the behavior?**

It should be possible to boot multiple angular 4 apps.

**Please tell us about your environment:**

I'm using webpack 2.3.3 with @ngtools/webpack loader (v. 1.2.13)

* **Angular version:** 4.0.1

* **Browser:** [all]

 
* **Language:** [all]

* **Node (for AoT issues):** `node --version` =   
node: 6.9.4
npm: 3.10.10

**Comment by tbosch** [comment]
I don't think this is an actual bug, but a feature:
If you are using the same component in multiple apps, then you only want to have the styles 1x in the browser, and not multiple times.

What are the downsides why this should be a bug?

**Comment by brechtbilliet** [comment]
is it a good idea to run multiple angular apps simultanously?
What happens when your DOM node gets removed? I don't see an "instance-specific" destroy function anywhere. Wouldn't you create memory leaks like this?

**Comment by tbosch** [comment]
`bootstrap` returns a `NgModuleRef` which has a `destroy` method.

**Comment by tbosch** [comment]
We do have an issue about memory leaking if you bootstrap multiple times though, see e.g. https://github.com/angular/angular/issues/13725.

So why is using the same encapsulation selector a bug in AOT? Do different component classes now share the same selector?

**Comment by d-bechtel** [comment]
the css class "app" is adopted to both apps, app1 and app2, which is a wrong behavior. If i load a second angular app it's style will overwrite the style of app1 due to the same css class.

**Comment by tytskyi** [comment]
@d-bechtel do we understand correct that 2 totally different applications with totally different components do generate the same css selectors because of incremental generation logic (root - c0, deep c1 etc)? 

So that totally different components' css are visually affecting each other?

**Comment by petersalomonsen** [comment]
According to the documentation for APP_ID ( https://angular.io/api/core/APP_ID ) it should be prefixing the css styles but I cannot see this happening. Also when looking at the appendix "Inspecting generated CSS" from https://angular.io/guide/component-styles where you have _ngcontent-pmm-5 (I guess pmm is the app id here), this is not the case with current angular where it would only be _ngcontent-5 (pmm is missing).

Shouldn't the APP_ID be part of the css prefix so that we could have multiple apps where styles are not overlapped?

This seems like a bug to me when looking at the documentation.

**Comment by petersalomonsen** [comment]
My workaround (quick fix) for preventing multiple apps overlapping css was to replace each occurrence of _nghost and _ngcontent with _nghost-myapp and _ngcontent-myapp in the generated aot.

Seems to work ok, but not an ideal solution.

Need to get the app id into the EmulatedEncapsulationDomRenderer2 class so that it can be passed to shimContentAttribute/shimHostAttribute along with the component id ( platform-browser/src/dom/dom_renderer.ts line 239 / 240 ).

**Comment by petersalomonsen** [comment]
Bug fix pull request:

#17745

**Comment by trotyl** [comment]
**This problem is nothing to do with JIT or AOT compilation**, but about whether multiple sets of Angular libraries being loaded at runtime, like:

```
<script src="app1/node_modules/@angular/core/bundles/core.js"></script>
<script src="app2/node_modules/@angular/core/bundles/core.js"></script>
<script src="app3/node_modules/@angular/core/bundles/core.js"></script>
```

Just informative here, this will only work within modules.

@d-bechtel It's very simple to reproduce in JIT, see http://embed.plnkr.co/2BY7dhyIQvSDF1RukMLl/

@petersalomonsen Your solution won't help with this problem, and the "problem" you were trying to solve does not exist from the beginning.

@tbosch I don't think this usage is ever officially supported in Angular, isn't it? But this scenario should still be valid especially for incoming Angular Elements, I'd suggest open a new feature request for that case. What do you think?

**Comment by trotyl** [comment]
Update: this is a regression caused by [`126fda2`](https://github.com/angular/angular/commit/126fda2613b1ad87ff62591bb1d3a23cd8f369a1#diff-d8d7af6e45117a338e7c3bdc10701ccfL39), landed in `4.0.0-rc.2`.

The `appId` was indeed used before that change:

https://github.com/angular/angular/blob/6b57928d35730c3833e62581960a474b644770cd/modules/%40angular/platform-browser/src/dom/dom_renderer.ts#L39-L40

**Comment by smasala** [comment]
Is there some way ugly way to overwrite the two functions `shimContentAttribute` and `shimHostAttribute`?

**Comment by glipecki** [comment]
As a workaround, we use post build script to change ngcontent/nghost names to ${appId} prefixed versions in target bundles. It's not elegant, but it works.

**Comment by smasala** [comment]
Below is the solution I went for in the end until this issue is hopefully resolved. It doesn't require any extra build processes except that it won't work with `buildOptimizer` active (`angular.json`) as it requires class meta data to find the correct instance.

https://github.com/smasala/angular-unique-shadow-dom/tree/master

**Comment by subashan** [comment]
Is this going to be fixed? We have an issue where we created a WebComponent using Angular Elements and it's interfering with our angular app.

**Comment by petersalomonsen** [comment]
A pull request with fix is here: https://github.com/angular/angular/pull/17745

It was made before the ivy renderer, so I haven't updated it for that - but still waiting for comments from the Angular team. (I'm happy to look into ivy support if they want this fix - but might be that they want to solve it another way).

**Comment by mleibman** [comment]
Friendly ping :)
We have some dependencies on this as well.

**Comment by petersalomonsen** [comment]
The pull request is merged: https://github.com/angular/angular/commit/ca14509bfed1766eab041726ecbe3675f8bda49e

**Comment by pkozlowski-opensource** [comment]
Sounds like this one was fixed long time ago by #17745
